File: core/src/nl/t64/game/rpg/screens/world/entity/PhysicsSparkle.java
package nl.t64.game.rpg.screens.world.entity;

import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.audio.AudioEvent;
import nl.t64.game.rpg.components.loot.Loot;
import nl.t64.game.rpg.screens.world.entity.events.Event;
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent;
import nl.t64.game.rpg.screens.world.entity.events.OnActionEvent;


public class PhysicsSparkle extends PhysicsComponent {

    private final Loot sparkle;
    private boolean isSelected;

    public PhysicsSparkle(Loot sparkle) {
        this.sparkle = sparkle;
        this.isSelected = false;
    }

    @Override
    public void receive(Event event) {
        if (event instanceof LoadEntityEvent loadEvent) {
            currentPosition = loadEvent.position;
        }
        if (event instanceof OnActionEvent onActionEvent) {
            if (onActionEvent.checkRect().overlaps(getRectangle())) {
                isSelected = true;
            }
        }
    }

    @Override
    public void update(Entity entity, float dt) {
        if (isSelected) {
            isSelected = false;
            Utils.getBrokerManager().componentObservers.notifyShowFindDialog(sparkle, AudioEvent.SE_SPARKLE);
        }
    }

    @Override
    public void debug(ShapeRenderer shapeRenderer) {
        // empty
    }

}

File: core/src/nl/t64/game/rpg/screens/world/entity/PhysicsSparkle.kt
package nl.t64.game.rpg.screens.world.entity

import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.components.loot.Loot
import nl.t64.game.rpg.screens.world.entity.events.Event
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent
import nl.t64.game.rpg.screens.world.entity.events.OnActionEvent


class PhysicsSparkle(private val sparkle: Loot) : PhysicsComponent() {

    private var isSelected = false

    override fun receive(event: Event) {
        if (event is LoadEntityEvent) {
            currentPosition = event.position
        }
        if (event is OnActionEvent) {
            if (event.checkRect.overlaps(getRectangle())) {
                isSelected = true
            }
        }
    }

    override fun update(entity: Entity, dt: Float) {
        if (isSelected) {
            isSelected = false
            brokerManager.componentObservers.notifyShowFindDialog(sparkle, AudioEvent.SE_SPARKLE)
        }
    }

    override fun debug(shapeRenderer: ShapeRenderer) {
        // empty
    }

}
